news/2017-05-19-tour-around-weld-3.asciidoc

```diff
@@ -0,0 +1,194 @@
+It follows a familiar builder pattern and the body of the observer methods is specified as lambda expression.
+Before booting Se container, we register this extension as we would any other.
+
+=== On-demand Request Context Activation
+
+Especially in SE (although not only there) you might want to activate a `RequestContext` manually for certain period of time.
+There are two ways to achieve that, first of which is an interceptor - `@ActivateRequestContext`.
+You can use that on either a method or a type (enabling it for all methods).
+As you might expect, it will activate the context before executing your method and shut it down afterwards.
+The other way is through means of built-in bean named `RequestContextController`.
+This bean can be injected as any other CDI bean and offers to self-explanatory methods: `activate` and `deactivate`.
+The obvious advantage of this approach is that you can enable the context for an extended period of time.
+
+=== Observer Method Ordering
+
+A small but noteworthy improvement to observer methods.
+You can not leverage `@Priority` annotation in observer methods hence ordering them as you wish.
+
+[ source, java ]
+----
+public void observeFirst (@Observes @Priority(1) EventPayload payload) {...}
+
+public void observeSecond (@Observes @Priority(2) EventPayload payload) {...}
+----
+
+=== Intercepting Produced Beans
+
+Up until now, anything you created using producers could not be easily intercepted.
+CDI 2.0 allows this for `@AroundInvoke` interceptors in two ways.
+There is a new method on `BeanManager` named `createInterceptionFactory` but most of the time you will rather use a built-in bean `InterceptionFactory` which can be injected as a producer method parameter.
+Here is a snippet:
+
+[ source, java ]
+----
+
+@Produces
+@Dependent
+public ProducedBean produceFoo(InterceptionFactory<ProducedBean> factory) {
+    factory.configure().add(Counter.Literal.INSTANCE);
+    return factory.createInterceptedInstance(new ProducedBean());
+}
+
+----
+
+To explain this a bit, we first inject the built-in bean with the type equal to the produced type we wish to intercept.
+Then we `configure()` it, which returns an `AnnotatedTypeConfigurator` allowing us to add interceptor binding.
+Finally, we invoke `InterceptionFactory.createInterceptedInstance()` which takes a the object we produce as a parameter.
+
+You can read more about this in link:http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#interception_factory[this CDI spec chapter, window="_blank"].
+
+=== Trimmed Bean Archives
+
+Last but not least feature we will mention as so called 'trimmed' bean archives.
+You can mark an explicit bean archive as trimmed in `beans.xml` by adding the `</trim>` element.
+Such bean archive will perform an annotated type discovery as with `bean-discovery-mode="all"` but all types that don't have a bean defining annotation are then removed from the set of discovered types.
```

Comment:
We should probably mention http://docs.jboss.org/weld/reference/latest-master/en-US/html/configure.html#veto-types-without-bean-defining-annotation and the difference...